README.md
# Mojo Package Index

![](mojopi_logo.jpg)


If something is not working for you, please check out [0.1.0](https://github.com/drunkwcodes/mojopi/tree/0.1.0) first for the smoothest experience.

`git checkout 0.1.0`

### Installation and Usage


- Use `pip install pdm`
- `git clone git@github.com:drunkwcodes/mojopi.git`
- `cd mojopi`
- `pdm install -d`

- activate venv

    Windows:
    `.venv\Scripts\activate`

    Linux:
    `.venv/bin/activate`

- `python src/mojopi/admin.py` to start server with admin mode.
- Open web browser, open 127.0.0.1:5000

### Screenshots

1. Index Page

![](screenshots/first_page.png)

2. Search Result

![](screenshots/search_result.png)

3. Project Page

![](screenshots/project_landing_page.png)

4. Project Releases

![](screenshots/releases.png)

5. Project Downloads

![](screenshots/downloads.png)

6. Login Page

![](screenshots/login.png)

7. Profile

![](screenshots/profile_page.png)


### API doc

open `http://127.0.0.1:5000/doc`, you will see:

![](screenshots/OpenAPI_doc.png)

![](screenshots/try_it.png)

### Contruibuting

Please see [CONTRIBUTING](CONTRIBUTING.md)

**Seeking experienced Materializecss frontend developers!**
There are quite a lot of UIs to be tweaked. The TODOs are in the [TODO.md](TODO.md)

CHANGELOG.md
# MojoPI 0.2.1 (2023-08-20)

### Bugfixes

- Fix duplicates in search result. Now there are only latest version project with corresponding names. (ba0e2d31)
- Use pdm's git dependecy and update REAME as well. (f553768b)


# MojoPI 0.2.0 (2023-08-03)

### Features

- OpenAPI doc is online now!
  Please open `127.0.0.1:5000/doc` to preview and play the API! (c5b443f2)


# MojoPI 0.1.0 (2023-08-01)

### Features

- Model Project established.
  add_project() function. **Thanks to PYPI.** (5f804115)
- User login, register, markdown profile, avatar. (7db1e85b)

### Bugfixes

- Move search bar to navbar successfully. (8a80ba20)
- Fix test. All test passed. But needs to be refined and calculate coverage. (#96917454)


test/unit/test_views.py
from flask import Flask

from mojopi.utils import login_manager
from mojopi.views import mbp


def test_mbp():
    app = Flask(__name__, template_folder="../../src/mojopi/templates")
    app.secret_key = "your_secret_key"
    login_manager.init_app(app)

    app.register_blueprint(mbp)

    client = app.test_client()

    response = client.get("/")
    assert response.status_code == 200

    response = client.get("/login")
    assert response.status_code == 200

    response = client.get("/register")
    assert response.status_code == 200

    response = client.get("/profile")
    assert response.status_code == 302

    response = client.get("/profile/1")
    assert response.status_code == 200

    response = client.get("/profile/100")
    assert response.status_code == 404

    # login simulation
    response = client.post(
        "/login", data=dict(email="eric@simutech.com.tw", password="123456")
    )
    assert response.status_code == 302

    with app.test_request_context():
        with client.session_transaction() as sess:
            pass  # TODO


test/unit/test_models.py
from pathlib import Path

from mojopi import InvalidInputError
from mojopi.models import User, add_user, init_db, is_valid_email, is_valid_username
from mojopi.utils import PIC_PATH, init_data


def test_add_user():
    init_data()
    init_db()
    try:
        add_user(
            email="eric@simutech.com.tw", username="drunkwcodes", password="123456"
        )
    except InvalidInputError:
        pass
    assert User.get(User.email == "eric@simutech.com.tw")


def test_is_valid_username():
    username = "drunkwcodes"
    u = User.get_or_none(User.username == username)
    if u:
        assert is_valid_username(username) is False
    else:
        assert is_valid_username(username) is True

    assert is_valid_username("") is False
    assert is_valid_username("../test") is False
    assert is_valid_username("stran_ge.username") is True


def test_is_valid_email():
    email = "eric@simutech.com.tw"
    if User.get_or_none(User.email == email):
        assert is_valid_email(email) is False
    else:
        assert is_valid_email(email) is True

    assert is_valid_email("") is False
    assert is_valid_email("asdf") is False


def test_change_username():
    username = "drunkwcodes"
    test_name = "testtest"
    u = User.get_or_none(User.username == username)
    if u:
        u.change_username(test_name)
        if u.picture:
            assert Path(u.picture).stem == test_name
            assert (Path(PIC_PATH) / u.picture).exists()

        u.change_username(username)


tests/unit/test_utils.py
import os
from pathlib import Path

from mojopi.models import User, add_user
from mojopi.utils import (
    DPATH,
    generate_password,
    hash_password,
    init_data,
    is_email,
    verify_password,
)


def test_init_data():
    init_data()
    assert Path(DPATH).exists()


def test_generate_password():
    pw = generate_password()
    assert len(pw) == 5


def test_hash_password():
    pw = hash_password(generate_password())
    assert len(pw) == 60


def test_verify_password():
    pw = generate_password()
    hpw = hash_password(pw)
    assert verify_password(pw, hpw)


def test_is_email():
    assert is_email("strange@email.com") is True
    assert is_email("str.ange@email.com") is True


scripts/muse.py
import json
from pathlib import Path
from pprint import pprint

import requests
from requests_toolbelt.multipart.encoder import MultipartEncoder
from resolvelib.resolvers import Resolver

# TODO: def parse_requirement(line: str, editable: bool = False) -> Requirement:
# src/pdm/models/requirements.py L480


def download():
    # a rough demo of downloading the mojo package.

    url = "http://127.0.0.1:5000/files/ring/test"
    r = requests.get(url)
    if r.status_code == 200:
        # print(f"{r.json() = }")  # no good.
        # print(f"{r.content = }")  # Oops.
        # print(f"{r.headers['Content-Disposition'].split('=')[-1] = }")  # Hmm.
        # pprint(r.headers["x-ring-info"])  # OUYA
        # print(f"{json.loads(r.headers['x-ring-info'])['file_name'] = }")
        file_name = r.headers["Content-Disposition"].split("=")[-1]
        with open(file_name, "wb") as f:
            f.write(r.content)
        print("file downloaded")
    else:
        print(f"{r.status_code = }")
        print(f"{r.text =}")


def upload():
    info = {"name": "test"}
    info["version"] = "3"
    info["author"] = "Eric"
    info["author_email"] = "eric@simutech.com.tw"
    info["requires_mojo"] = "0.1"
    info["file_name"] = "test-3.ring"
    info["description"] = "**Hellog mojopi**"
    info["description_content_type"] = "text/markdown"
    info["home_page"] = "http://github.com/drunkwcodes"
    info["keywords"] = "test, mojopi"
    info["license"] = "MIT"
    info["maintainer"] = "drunkwcodes"
    info["maintainer_email"] = "drunkwcodes@gmail.com"
    info["summary"] = "test summary"

    url = "http://127.0.0.1:5000/api/ring/test/3"

    file_path = Path(__file__).with_name("test-3.ring")

    # 使用 MultipartEncoder 來建構 multipart/form-data 的內容
    multipart_data = MultipartEncoder(
        fields={
            "file": (
                info["file_name"],
                open(file_path, "rb"),
                "application/octet-stream",
            ),
            "info": ("info.json", json.dumps(info), "application/json"),
        }
    )

    headers = {
        "x-ring-info": json.dumps(info),
        "Content-Type": multipart_data.content_type,
    }

    response = requests.post(url, data=multipart_data, headers=headers)

    # 檢查回應狀態碼
    if response.status_code == 200:
        print("File upload succeeded!")
        print(response.json())  # 如果 API 回傳 JSON 資料，可以用 .json() 解析回傳內容
    else:
        print(f"File upload failed with status code: {response.status_code}")
        print(response.json())  # 顯示回應內容 (通常會包含錯誤訊息)


if __name__ == "__main__":
    download()
    upload()


src/mojopi/app.py
import logging
from pathlib import Path

from flask import Flask

from mojopi.models import db_path, init_db
from mojopi.utils import DPATH, conf, init_data, login_manager
from mojopi.views import api, fbp, mbp, mojobp


def main():
    init_data()

    app = Flask(__name__)
    app.config["SECRET_KEY"] = "123456790"  # 設置一個密鑰以進行會話加密

    # setup logging
    log_file = DPATH / "mojopi.log"
    logging.basicConfig(filename=log_file, filemode="a")
    logging.getLogger().setLevel(logging.INFO)

    # setup flask login
    login_manager.init_app(app)

    # setup db
    if not Path(db_path).exists():
        init_db()

    # views
    app.register_blueprint(mbp)
    app.register_blueprint(fbp)
    app.register_blueprint(mojobp)

    # init api
    api.init_app(app)

    app.run()


if __name__ == "__main__":
    main()


src/mojopi/admin.py
import logging
from pathlib import Path

import flask_admin as fadmin
from flask import Blueprint, Flask, render_template, url_for
from flask_admin.contrib.peewee import ModelView
from flask_login import current_user

from mojopi.models import User, db_path, init_db
from mojopi.utils import DPATH, conf, init_data, login_manager
from mojopi.views import api, fbp, mbp, mojobp


class UserAdmin(ModelView):
    column_display_pk = True


def main():
    init_data(mock=conf.get("mock_data", True))

    app = Flask(__name__)
    app.config["SECRET_KEY"] = "123456790"  # 設置一個密鑰以進行會話加密

    # setup logging
    log_file = DPATH / "mojopi.log"
    logging.basicConfig(filename=log_file, filemode="a")
    logging.getLogger().setLevel(logging.DEBUG)

    # setup flask login
    login_manager.init_app(app)

    # setup flask admin
    admin = fadmin.Admin(app, name="MojoPI Admin")

    admin.add_view(UserAdmin(User))

    # setup db
    if not Path(db_path).exists():
        init_db(mock=conf.get("mock_db", True))

    # views
    adminbp = Blueprint("adminbp", __name__)

    @adminbp.route("/")
    def index():
        if current_user.is_authenticated:
            return render_template("index.html", admin_mode=True)
        return render_template("index.html", admin_mode=True)

    app.register_blueprint(
        adminbp
    )  # the router will route to the first registered view.
    app.register_blueprint(mbp)
    app.register_blueprint(fbp)
    app.register_blueprint(mojobp)

    # init api
    api.init_app(app)

    app.run(debug=True)


if __name__ == "__main__":
    main()


src/mojopi/utils.py
"""一些基本的函數，不牽涉到其他模組的東西。盡量不要 import other project's module to avoid circular imports."""

import os
import random
import re
import string
from hashlib import sha256

try:
    import tomllib
except ImportError:
    import tomlkit

from pathlib import Path

import bcrypt
from flask_login import LoginManager

conf_file = Path(__file__).with_name("conf.toml")
try:
    with open(conf_file, "rb") as f:
        conf = tomllib.load(f)
except NameError:
    with open(conf_file, "r") as f:
        conf = tomlkit.load(f)

login_manager = LoginManager()


class InvalidInputError(Exception):
    pass


if (
    conf.get("data_folder") is None or not Path(conf.get("data_folder")).parent.exists()
):  # The folder contains data not exists
    if "python" in str(Path(__file__)).lower():
        DPATH = Path(os.getcwd()) / "data"  # TODO: test
    else:
        DPATH = Path(__file__).parent.parent.parent / "data"
else:
    DPATH = Path(conf["data_folder"])


SERVER_FILES_PATH = DPATH / "server_files"
PIC_PATH = DPATH / "server_files" / "profile_pic"
RINGS_PATH = DPATH / "server_files" / "rings"
PUBLIC_PATH = DPATH / "server_files" / "public"


MOCK_DB = conf.get("mock_db", False)
MOCK_DATA = conf.get("mock_data", False)


def init_data(mock=MOCK_DATA):
    if not os.path.exists(Path(DPATH)):
        os.makedirs(DPATH)

    # mkdir fbp
    if not os.path.exists(SERVER_FILES_PATH):
        os.mkdir(SERVER_FILES_PATH)

    # mkdir profile picture
    if not os.path.exists(PIC_PATH):
        os.mkdir(PIC_PATH)

    # mkdir ring files
    if not os.path.exists(RINGS_PATH):
        os.mkdir(RINGS_PATH)

    # public server files
    if not PUBLIC_PATH.exists():
        os.mkdir(PUBLIC_PATH)

    if not mock:
        return

    # start mocking
    # create mock ring
    with open(RINGS_PATH / "test-2.ring", "w", encoding="utf-8") as f:
        f.writelines("123")


def generate_password(length=5):
    clist = string.ascii_letters + string.digits
    pw = []
    for _ in range(length):
        pw.append(random.choice(clist))
    return "".join(pw)


def hash_password(password):
    # Generate a salt and hash the password
    salt = bcrypt.gensalt()
    hashed_password = bcrypt.hashpw(password.encode("utf-8"), salt)

    return hashed_password


def verify_password(password, hashed_password):
    # Check if the provided password matches the hashed password
    if password is None:
        password = ""
    if hashed_password is None:
        hashed_password = ""

    if not isinstance(password, bytes):
        password = password.encode("utf-8")

    if not isinstance(hashed_password, bytes):
        hashed_password = hashed_password.encode("utf-8")

    return bcrypt.checkpw(password, hashed_password)


def is_email(email):
    pattern = r"^[\w\.-]+@[\w\.-]+\.\w+$"  # 正則表達式模式
    return re.match(pattern, email) is not None


def file_size(file_path):
    try:
        size = os.path.getsize(file_path)
        return size
    except FileNotFoundError:
        print(f"file not found：{file_path}")
    except Exception as e:
        print(f"error：{e}")
        return None


def hr_size(size_in_bytes):
    """
    Calculate the human-readable size of a file.

    Args:
        size_in_bytes (int): The size of the file in bytes.

    Returns:
        str: The human-readable size of the file.
    """
    # 定義不同的大小單位
    units = ["B", "KB", "MB", "GB", "TB"]

    # 若檔案大小小於 1 B，直接回傳
    if size_in_bytes < 1:
        return f"{size_in_bytes} B"

    # 計算使用哪個單位
    unit_index = 0
    while size_in_bytes >= 1024 and unit_index < len(units) - 1:
        size_in_bytes /= 1024
        unit_index += 1

    # 使用適當的單位進行格式化
    return f"{size_in_bytes:.2f} {units[unit_index]}"


def calculate_sha256(file_path):
    sha256_hash = sha256()

    # 使用二進位模式打開檔案並將整個內容讀取到記憶體中
    with open(file_path, "rb") as file:
        file_content = file.read()

    # 將檔案內容傳遞給 hashlib.sha256 函數
    sha256_hash.update(file_content)

    # 回傳計算得到的 SHA-256 雜湊值（以十六進位字串表示）
    return sha256_hash.hexdigest()


src/mojopi/views.py
import json
import logging
from dataclasses import dataclass, is_dataclass
from pathlib import Path

from flask import (
    Blueprint,
    abort,
    flash,
    jsonify,
    make_response,
    redirect,
    render_template,
    request,
    send_file,
    send_from_directory,
    url_for,
)
from flask_login import current_user, login_user, logout_user
from flask_restx import Api, Resource, fields
from mups import ring_info
from mups.core import RingInfo
from peewee import DoesNotExist, IntegrityError, fn
from werkzeug.utils import secure_filename

from . import __version__
from .models import (
    Profile,
    Project,
    Ring,
    User,
    add_project,
    add_ring,
    add_user,
    is_valid_email,
    is_valid_username,
)
from .readme import render  # TODO: Not working, need to investigate
from .utils import (
    DPATH,
    PIC_PATH,
    RINGS_PATH,
    SERVER_FILES_PATH,
    InvalidInputError,
    calculate_sha256,
    hash_password,
    login_manager,
    verify_password,
)

# apibp = Blueprint("apibp", __name__, url_prefix="/api")  # json api

mbp = Blueprint(
    "mbp", __name__
)  # main bp, including index page, login, register, profile, settings etc.

mojobp = Blueprint("mojobp", __name__)  # everything related to mojo

# TODO: 404 handler. PYPI has a search bar and good color.

# files bp
fbp = Blueprint(
    "fbp",
    __name__,
    static_folder=SERVER_FILES_PATH / "public",
    static_url_path="/public",
    url_prefix="/files",
)


# 對應不同型別和 fields 類之間的對應關係
TYPE_TO_FIELD = {
    int: fields.Integer,
    str: fields.String,
    bool: fields.Boolean,
    float: fields.Float,
    list: fields.List,
    # 在這裡添加其他型別對應的 fields 類
}


# 轉換 dataclass 屬性到資料模型的 fields
def convert_to_fields(data_class):
    fields_dict = {}
    for field_name, field_type in data_class.__annotations__.items():
        if is_dataclass(field_type):
            fields_dict[field_name] = convert_to_fields(field_type)
        else:
            field_class = TYPE_TO_FIELD.get(field_type, fields.Raw)
            fields_dict[field_name] = field_class(description=f"The {field_name} field")
    return fields_dict


# 初始化資料模型
api = Api(
    version=__version__,
    title="MojoPI API",
    description="MojoPI API are all list here",
    doc="/doc",
    terms_url="https://github.com/drunkwcodes/mojopi/blob/main/terms_of_service.txt",
    contact="drunkwcodes@gmail.com",
    license="MIT",
    license_url="https://github.com/drunkwcodes/mojopi/blob/main/LICENSE",
    prefix="/api",
)

ring_info_model = api.model("RingInfoModel", convert_to_fields(RingInfo))


def profile_pic_url(user_id=0, username=""):
    # default anonymous avatar
    url = url_for("static", filename="Sample_User_Icon.png")

    if user_id:
        try:
            user = User.get_by_id(user_id)
        except DoesNotExist:
            return url
    elif username:
        user = User.get_or_none(User.username == username)
        if user is None:
            return url
    else:
        return url

    if user.picture:
        user_id = user.id
        url = url_for("fbp.get_profile_pic", user_id=user_id)

    return url


@mbp.route("/")
def index():
    if current_user.is_authenticated:
        return render_template("index.html")
    return render_template("index.html")


@login_manager.user_loader
def load_user(user_id):
    try:
        return User.get_by_id(user_id)
    except DoesNotExist:
        return None


@mbp.route("/login", methods=["GET", "POST"])
def login():
    if request.method == "GET":
        next_url = request.args.get("next")
        return render_template("login.html", next_url=next_url)

    elif request.method == "POST":
        email = request.form.get("email")  # form 用 name="email" 才抓得到
        password = request.form.get("password")
        next_url = request.args.get("next")
        user = User.get_or_none(User.email == email)

        if user:
            if not user.password:
                login_user(user)
                flash("Need to reset password.")
                return redirect(next_url or url_for("mbp.reset_password"))

            elif verify_password(password=password, hashed_password=user.password):
                login_user(user)
                flash("Logged in successfully.")

                return redirect(next_url or url_for("mbp.index"))
            else:
                flash("Wrong password.")
                return redirect(url_for("mbp.login", next=next_url))
        else:
            flash("Email incorrect.")
            return redirect(url_for("mbp.login"))  # TODO: 或是顯示錯誤訊息


@mbp.route("/register", methods=["GET", "POST"])
def register():
    if request.method == "GET":
        return render_template("register.html")

    elif request.method == "POST":
        email = request.form.get("email")
        username = request.form.get("username")
        password = request.form.get("password")

        if not is_valid_email(email) or not is_valid_username(username):
            flash("Register failed. Invalid email and username.")
            return redirect(url_for("mbp.index"))

        # start registering
        try:
            user, _ = add_user(email=email, username=username, password=password)
        except InvalidInputError:
            flash("Register failed. Invalid email and username.")
            return redirect(url_for("mbp.index"))

        login_user(user)
        flash("Register successfully.")

        return redirect(url_for("mbp.index"))


@api.route("/username/<string:usn>")
class UsernameResource(Resource):
    def get(self, usn):
        if is_valid_username(usn):
            return {"valid": True}
        else:
            return {"valid": False}


@api.route("/email/<string:eml>")
class EmailResource(Resource):
    def get(self, eml):
        if is_valid_email(eml):
            return {"valid": True}
        else:
            return {"valid": False}


@mbp.route("/profile")
def profile_self():
    if not current_user.is_authenticated:
        return redirect(url_for("mbp.login", next="/profile"))
    return redirect(f"/profile/{current_user.id}")


@mbp.route("/profile/<int:user_id>")
def profile(user_id):
    try:
        user = User.get_by_id(user_id)
    except DoesNotExist:
        abort(404)

    pf = Profile.get(Profile.user == user)
    if pf.is_public or current_user == user:
        return render_template(
            "profile.html",
            profile_pic_url=profile_pic_url(user_id=user_id),
            profile=pf,
        )
    else:
        abort(404, "User's profile is private.")


@fbp.route("/profile_pic", methods=["POST"])
def profile_pic_upload():
    if not current_user.is_authenticated:
        abort(401)

    if "file" not in request.files:
        abort(400, "No file provided.")

    file = request.files["file"]
    if file.filename == "":
        abort(400, "Empty filename.")

    if file:
        # 使用 secure_filename 函式來避免潛在的安全問題
        filename = secure_filename(file.filename)
        username = current_user.username
        filename_suffix = Path(filename).suffix  # 取得檔案副檔名

        # 儲存檔案到指定的目錄中
        file.save(PIC_PATH / f"{username}{filename_suffix}")

        current_user.picture = f"{username}{filename_suffix}"
        current_user.save()

        return {"message": "Profile picture saved successfully."}
    else:
        abort(400, "Invalid file.")


@fbp.route("/profile_pic", methods=["GET"], defaults={"user_id": None})
@fbp.route("/profile_pic/<int:user_id>", methods=["GET"])
def get_profile_pic(user_id):
    if user_id is None:
        if not current_user.is_authenticated:
            abort(401)
        user = current_user
    else:
        try:
            user = User.get_by_id(user_id)
        except DoesNotExist:
            abort(404, description="User not found.")

    if user.picture:
        pic_path = PIC_PATH / user.picture
        if pic_path.is_file():
            return send_from_directory(str(pic_path.parent), pic_path.name)
    abort(404, "Profile picture not found.")


@mbp.route("/edit_profile", methods=["GET", "POST"])
def edit_profile():
    if not current_user.is_authenticated:
        abort(401)

    if request.method == "GET":
        pf = Profile.get(Profile.user == current_user)
        return render_template("edit_profile.html", profile=pf)

    elif request.method == "POST":
        username = request.form.get("username")
        education = request.form.get("education")
        experience = request.form.get("experience")
        bio = request.form.get("bio")

        current_user.username = username
        try:
            current_user.save()
        except IntegrityError:
            flash("New username conflicts with existing account.")

        pf = Profile.get(Profile.user == current_user)
        pf.education = education
        pf.experience = experience
        pf.bio = bio
        pf.save()

        flash("Edit Successful!")
        return redirect(url_for("mbp.profile", user_id=current_user.id))


@mbp.route("/settings")
def settings():
    if not current_user.is_authenticated:
        abort(401, description="Need to login to view settings.")
    return render_template("settings.html")


@mbp.route("/reset_password", methods=["GET", "POST"])
def reset_password():
    if not current_user.is_authenticated:
        abort(401)

    if request.method == "GET":
        return render_template("reset_password.html")

    elif request.method == "POST":
        new_password = request.form.get("new-password")
        confirm_password = request.form.get("confirm-password")

        try:
            assert new_password == confirm_password
        except AssertionError:
            abort(403, "New password is not equal to confirm password.")

        current_user.password = hash_password(new_password)
        current_user.save()

        flash("New password set!")

        return redirect(url_for("mbp.settings"))


@mbp.route("/logout")
def logout():
    logout_user()
    return redirect(url_for("mbp.index"))


def project_info(pname, version=""):
    pj = Project.get_or_none((Project.name == pname) & (Project.version == version))
    latest_ring_dt = (
        Ring.select(fn.MAX(Ring.upload_at)).where(Ring.name == pname).scalar()
    )
    ring = Ring.get_or_none((Ring.name == pname) & (Ring.upload_at == latest_ring_dt))

    if pj is None:
        return None

    info = {"name": pname}
    info["version"] = pj.version
    if pj.description:
        info["description"] = pj.description
    if pj.description_content_type:
        info["description_content_type"] = pj.description_content_type
    if pj.home_page:
        info["home_page"] = pj.home_page
    if pj.keywords:
        info["keywords"] = pj.keywords
    if pj.license:
        info["license"] = pj.license
    if ring is not None:
        if ring.author:
            info["author"] = ring.author
        if ring.author_email:
            info["author_email"] = ring.author_email
        if ring.requires_mojo:
            info["requires_mojo"] = ring.requires_mojo
        if ring.file_name:
            info["file_name"] = ring.file_name

    if pj.maintainer:
        info["maintainer"] = pj.maintainer
    if pj.maintainer_email:
        info["maintainer_email"] = pj.maintainer_email
    if pj.summary:
        info["summary"] = pj.summary
    info["create_at"] = str(pj.create_at)
    info["last_modified"] = str(pj.last_modified)

    return info


def composite_info(ring):
    pj = Project.get_or_none(
        (Project.name == ring.name) & (Project.version == ring.version)
    )

    info = {"name": ring.name}
    info["version"] = ring.version
    if ring.author:
        info["author"] = ring.author
    if ring.author_email:
        info["author_email"] = ring.author_email
    if ring.requires_mojo:
        info["requires_mojo"] = ring.requires_mojo
    if ring.file_name:
        info["file_name"] = ring.file_name

    if pj is not None:
        if pj.description:
            info["description"] = pj.description
        if pj.description_content_type:
            info["description_content_type"] = pj.description_content_type
        if pj.home_page:
            info["home_page"] = pj.home_page
        if pj.keywords:
            info["keywords"] = pj.keywords
        if pj.license:
            info["license"] = pj.license
        if pj.maintainer:
            info["maintainer"] = pj.maintainer
        if pj.maintainer_email:
            info["maintainer_email"] = pj.maintainer_email
        if pj.summary:
            info["summary"] = pj.summary
        info["create_at"] = str(pj.create_at)
        info["last_modified"] = str(pj.last_modified)

    return info


def latest_project_version(project_name):
    latest_version = (
        Project.select(fn.MAX(Project.version))
        .where(Project.name == project_name)
        .scalar()
    )
    return latest_version


@mojobp.route("/project/<string:project_name>", defaults={"version": ""})
@mojobp.route("/project/<string:project_name>/<string:version>")
def project(project_name, version):
    if version == "":
        version = latest_project_version(project_name=project_name)

    pj = Project.get_or_none(
        (Project.name == project_name) & (Project.version == version)
    )
    if pj is None:
        abort(404)
    return render_template(
        "project.html",
        info=project_info(project_name, version=version),
        description=render(pj.description, content_type=pj.description_content_type),
    )


@mojobp.route("/project/<string:project_name>/history", defaults={"version": ""})
@mojobp.route("/project/<string:project_name>/<string:version>/history")
def project_history(project_name, version):
    if version == "":
        version = latest_project_version(project_name=project_name)

    pj = Project.get_or_none(
        (Project.name == project_name) & (Project.version == version)
    )
    if pj is None:
        abort(404, "Project not exist.")

    releases = (
        Ring.select().where(Ring.name == project_name).order_by(Ring.upload_at.desc())
    )

    return render_template(
        "project_history.html",
        info=project_info(project_name, version=version),
        releases=releases,
    )


@mojobp.route("/project/<string:project_name>/files", defaults={"version": ""})
@mojobp.route("/project/<string:project_name>/<string:version>/files")
def project_files(project_name, version):
    if version == "":
        version = latest_project_version(project_name=project_name)

    pj = Project.get_or_none(
        (Project.name == project_name) & (Project.version == version)
    )
    if pj is None:
        abort(404, "Project not exist.")

    rings = Ring.select().where((Ring.name == project_name) & (Ring.version == version))

    # TODO: source dist
    return render_template(
        "project_download.html",
        info=project_info(project_name, version=version),
        rings=rings,
    )


@fbp.route(
    "/ring/<string:name>",
    defaults={"version": "", "platform": ""},
    methods=["GET", "POST"],
)
@fbp.route(
    "/ring/<string:name>/<string:version>",
    defaults={"platform": ""},
    methods=["GET", "POST"],
)
@fbp.route(
    "/ring/<string:name>/<string:version>/<string:platform>", methods=["GET", "POST"]
)
def ring_file(name, version, platform):  # TODO: Add upload test script
    if not version:
        version = latest_project_version(name)

    if request.method == "GET":
        if platform:
            ring = Ring.get_or_none(
                (Ring.name == name)
                & (Ring.version == version)
                & (Ring.platform == platform)
            )
        else:  # TODO: determine a default platform
            ring = Ring.get_or_none((Ring.name == name) & (Ring.version == version))

        if ring is None:
            abort(
                404, f"Ring not found: {name}, version: {version}, platform: {platform}"
            )

        if ring.file_name:
            # Your logic to generate the additional info JSON
            additional_info = composite_info(ring)

            # Sending the file from the specified directory
            file_path = RINGS_PATH / ring.file_name

            # Creating a custom response with file content and additional info
            response = make_response(send_file(file_path))

            # Adding additional_info to the response headers
            response.headers["x-ring-info"] = json.dumps(additional_info)

            return response

        else:
            abort(404, "Ring file not uploaded.")

    elif request.method == "POST":  # TODO: dead code. To be removed.
        if not request.is_json:
            abort(415, "Not JSON")

        file = request.files["file"]
        if file.filename == "":
            abort(400, "No file upload.")

        data = request.get_json()
        name = data.get("name")
        version = data.get("version", "")
        platform = data.get("platform", "")
        author = data.get("author", "")
        author_email = data.get("author_email", "")
        require_dist = data.get("require_dist", "")
        require_mojo = data.get("require_mojo", "")
        file_name = file.filename
        sha256 = calculate_sha256(RINGS_PATH / file_name)

        if name is None:
            abort(
                422, "name not provided."
            )  # https://developer.mozilla.org/zh-CN/docs/Web/HTTP/Status/422

        try:
            add_ring(
                name=name,
                version=version,
                platform=platform,
                author=author,
                author_email=author_email,
                require_dist=require_dist,
                require_mojo=require_mojo,
                file_name=file_name,
                sha256=sha256,
            )
        except InvalidInputError:
            abort(400, "Upload failed. Duplicate ring.")

        file.save(RINGS_PATH / file.filename)


@mojobp.route("/search")
def search():
    # 從 URL 中取得搜尋字串
    query_string = request.args.get("q", "")

    # 在資料庫  或內容中進行搜尋操作

    pjs = Project.select().where(Project.name ** f"%{query_string}%")
    results = []
    pj_names = {pj.name for pj in pjs}
    for pj_name in pj_names:
        results.append(
            Project.get_or_none(
                (Project.name == pj_name)
                & (Project.version == latest_project_version(pj_name))
            )
        )

    # 返回搜尋結果給前端頁面
    return render_template("search_results.html", results=results)


@api.route("/ring/<string:name>/<string:version>/<string:platform>")
@api.route("/ring/<string:name>/<string:version>")
class RingResource(Resource):
    def get(self, name, version, platform=""):
        ring = Ring.get_or_none(
            (Ring.name == name)
            & (Ring.version == version)
            & (Ring.platform == platform)
        )
        if ring is None:
            return {"message": "Ring not found."}, 404

        return composite_info(ring)

    def post(self, name, version, platform=""):
        ring = Ring.get_or_none(
            (Ring.name == name)
            & (Ring.version == version)
            & (Ring.platform == platform)
        )
        if ring is not None:
            print(f"{ring = }")
            return {"message": "Ring exists."}, 400

        # 檢查是否有檔案和 info 資料
        if "file" not in request.files or "info" not in request.files:
            return {"error": "Missing file or info data."}, 400

        # 取得檔案和 info 資料
        file = request.files["file"]
        info = request.files["info"].read()

        # 解析 info 資料
        try:
            info = json.loads(info)
        except json.JSONDecodeError:
            return {"error": "Invalid JSON format in info data."}, 400

        # 取得 info 中的檔案名稱
        file_name = info.get("file_name")

        # 確認檔案名稱存在且非空
        if not file_name:
            return {"error": "File name missing in info data."}, 400

        # 儲存檔案
        file.save(RINGS_PATH / file_name)
        add_ring(
            name=name,
            version=version,
            platform=platform,
            file_name=file_name,
            author=info.get("author"),
            author_email=info.get("author_email"),
            require_dist=info.get("require_dist"),
            requires_mojo=info.get("require_mojo"),  # TODO: rename require_mojo
            sha256=calculate_sha256(RINGS_PATH / file_name),
        )
        # 回傳成功訊息
        return {"message": "File uploaded successfully."}, 200


@api.route("/ring/<string:name>/<string:version>/download")
@api.route("/ring/<string:name>/<string:version>/<string:platform>/download")
class RingDownloadResource(Resource):
    def get(self, name, version, platform=""):
        ring = Ring.get_or_none(
            (Ring.name == name)
            & (Ring.version == version)
            & (Ring.platform == platform)
        )
        if ring is None:
            return {"message": "Ring not found."}, 404

        return redirect(
            url_for("fbp.ring_file", name=name, version=version, platform=platform)
        )
